### PR TITLE
Add common base exception to all p2p exceptions

### PR DESCRIPTION
**Pull request by cburgdorf**
### What was wrong?

As described in #591 the exceptions in p2p did not have a common project base exception yet.

### How was it fixed?

Added base exceptions and also added docstrings.

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

File: p2p/exceptions.py [hunk 1]
```diff
@@ -1,54 +1,95 @@
-class AuthenticationError(Exception):
+class BaseP2PError(Exception):
+    """
+    Base error class for all p2p errors.
+    """
     pass
 
-
-class DecryptionError(Exception):
+class AuthenticationError(BaseP2PError):
```
Comment (cburgdorf) [inline p2p/exceptions.py]: Judging from the callsites it seemed to me as if this and `DecryptionError` could get merged into one. Thoughts?

File: p2p/exceptions.py [hunk 2]
```diff
@@ -1,54 +1,95 @@
...
 
-class TooManyTimeouts(Exception):
+class LESAnnouncementProcessingError(BaseP2PError):
+    """
+    Error signaling that an LESAnnouncement could not be processed.
+    """
     pass
 
 
-class PeerFinished(Exception):
```
Comment (cburgdorf) [inline p2p/exceptions.py]: That one wasn't used anywhere

File: p2p/exceptions.py (hunk 1 above)
Comment (gsalgado) [reply p2p/exceptions.py]: Seems like it

Comment (cburgdorf) [reply p2p/exceptions.py]: Alright, will update the PR.

File: p2p/exceptions.py [hunk 3]
```diff
@@ -1,54 +1,89 @@
...
+    """
+    Base error class for all p2p errors.
+    """
     pass
 
 
-class DecryptionError(Exception):
+class DecryptionError(BaseP2PError):
+    """
+    Error signaling that a message could not be decrypted.
```
Comment (gsalgado) [inline p2p/exceptions.py]: I'd get rid of the "Error signaling that" part from all docstrings here.

File: p2p/peer.py [hunk 4]
```diff
@@ -396,7 +396,7 @@ def decrypt_header(self, data: bytes) -> bytes:
         self.ingress_mac.update(sxor(aes, header_ciphertext))
         expected_header_mac = self.ingress_mac.digest()[:HEADER_LEN]
         if not bytes_eq(expected_header_mac, header_mac):
-            raise AuthenticationError('Invalid header mac')
+            raise DecryptionError('Invalid header mac')
```
Comment (pipermerriam) [inline p2p/peer.py]: Curious why the exception class here was changed?

Comment (cburgdorf) [reply p2p/peer.py]: https://github.com/ethereum/py-evm/pull/602#discussion_r184365395

Both seemed to mean the same thing so I got rid of one.

**Comment by cburgdorf** [comment]
In regard to [the discussion](https://github.com/ethereum/py-evm/pull/602#discussion_r184398587) around the docstrings, I looked [how the standard lib handles this](https://docs.python.org/3/library/exceptions.html) and changed all docstrings to be aligned with that.

So, the pythonic way seems to be:

1. Base exceptions start with

>The base class for...

2. Other exceptions start with

>Raised when...

**Comment by pipermerriam** [comment]
👍 I love borrowing conventions from elsewhere.

**Comment by cburgdorf** [comment]
@pipermerriam This failed CI run seems like...not legit. Is that a bug in the Circle setup or just a hiccup and we need to retrigger (I can't)

**Pull request merged**